module/zfs/txg.c

```diff
@@ -791,24 +785,22 @@ txg_wait_open(dsl_pool_t *dp, uint64_t txg, boolean_t should_quiesce)
 }
 
 /*
- * If there isn't a txg syncing or in the pipeline, push another txg through
- * the pipeline by quiescing the open txg.
+ * Pass in the txg number that should be synced.
  */
 void
-txg_kick(dsl_pool_t *dp)
+txg_kick(dsl_pool_t *dp, uint64_t txg)
 {
 	tx_state_t *tx = &dp->dp_tx;
 
 	ASSERT(!dsl_pool_config_held(dp));
 
+	if (tx->tx_sync_txg_waiting >= txg)
+		return;
+
 	mutex_enter(&tx->tx_sync_lock);
-	if (!txg_is_syncing(dp) &&
-	    !txg_is_quiescing(dp) &&
-	    tx->tx_quiesce_txg_waiting <= tx->tx_open_txg &&
-	    tx->tx_sync_txg_waiting <= tx->tx_synced_txg &&
-	    tx->tx_quiesced_txg <= tx->tx_synced_txg) {
-		tx->tx_quiesce_txg_waiting = tx->tx_open_txg + 1;
-		cv_broadcast(&tx->tx_quiesce_more_cv);
+	if (tx->tx_sync_txg_waiting < txg) {
+		tx->tx_sync_txg_waiting = txg;
+		cv_broadcast(&tx->tx_sync_more_cv);
```

Comment:
As I understand, the idea behind separate quiescing and syncing threads was to allow them run in parallel.  With this change if syncing thread is busy, quiescing won't start, that looks to me like they will be be strictly serialized, i.e. pointless to me.

Quiescing has its role when some opened transaction can not be committed quickly, delaying all the group.  I think it was much more often before in write path dmu_tx_check_ioerr() started to prefetch all indirect blocks.  I think we need to make sure there is no other realistic reasons (memory allocations, swap-in's, etc.) for transactions to be delayed open before we break the quiescing

Reply:
In the original txg_kick(), quiescing and syncing can't run parallel, either.

I have looked into preemptive quiescing a txg when there is one still syncing #11929. I could not find any evidence that may improve measurable performance in that pull request. That's the part of reason I end up in this pull request.

Reply:
> In the original txg_kick(), quiescing and syncing can't run parallel, either.

They can, but as you correctly noticed quiescing usually completes too fast to be measurable.  What I am objecting here is breaking existing design without removing all the remnants.  Either we keep quiescing state, and then it must be functional, or we are absolutely sure we don't need it, and then we drop from 3 active txgs to only two.

Reply:
I was probably too aggressive about two txgs, it is slightly from different area, 3 would still be needed.  But dependent quiescing thread is definitely useless.

Reply:
Hi, @amotin 
I just pushed a commit, in which I switched back to tx_quiesce_more_cv for kick. It should be **equivalent** to the txg_kick() in current master of "openzfs/zfs". Also there is some improvement comparing with the current master.